sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/test/scala/com/azure/cosmos/spark/PartitionMetadataCacheSpec.scala

```diff
@@ -0,0 +1,169 @@
+    val initialLastUpdated = newItem.lastUpdated.get
+
+    //scalastyle:off magic.number
+    PartitionMetadataCache.applyTestOverrides(
+      newRefreshIntervalInMsOverride = Some(10),
+      newStaleCachedItemRefreshPeriodInMsOverride = Some(10 * 1000),
+      newCachedItemTtlInMsOverride = Some(50))
+
+    Thread.sleep(500)
+    //scalastyle:on magic.number
+
+    PartitionMetadataCache.purge(containerConfig, feedRange) shouldEqual false
+    this.reinitialize()
+  }
+
+  //scalastyle:off multiple.string.literals
+  it should "automatically update the cached item after staleness threshold elapsed" taggedAs RequiresCosmosEndpoint in {
+    this.reinitialize()
+
+    val newItem = PartitionMetadataCache(clientConfig, None, containerConfig, feedRange).block()
+    val initialLastUpdated = newItem.lastUpdated.get
+
+    //scalastyle:off magic.number
+    PartitionMetadataCache.applyTestOverrides(
+      newRefreshIntervalInMsOverride = Some(10),
+      newStaleCachedItemRefreshPeriodInMsOverride = Some(50),
+      newCachedItemTtlInMsOverride = Some(10 * 1000))
+
+    Thread.sleep(500)
+    //scalastyle:on magic.number
+
+    val candidate = PartitionMetadataCache(clientConfig, None, containerConfig, feedRange).block()
+    candidate.lastUpdated.get should be > initialLastUpdated
+
+    PartitionMetadataCache.purge(containerConfig, feedRange) shouldEqual true
+    this.reinitialize()
+  }
+
+  it should "honor test data overrides" taggedAs RequiresCosmosEndpoint in {
+    this.reinitialize()
+    val startEpochMs = Instant.now.toEpochMilli
+
+    val rnd = scala.util.Random
+    val docCount = rnd.nextInt()
+    val docSize = rnd.nextInt()
+    val lastLsn = rnd.nextInt()
+
+    val testMetadata = PartitionMetadata(
+      clientConfig,
+      None,
+      containerConfig,
+      feedRange,
+      docCount,
+      docSize,
+      lastLsn,
+      new AtomicLong(startEpochMs),
+      new AtomicLong(startEpochMs))
+
+    PartitionMetadataCache.injectTestData(containerConfig, feedRange, testMetadata)
+
+    //scalastyle:off magic.number
+    Thread.sleep(10)
+    //scalastyle:on magic.number
+
+    val newItem = PartitionMetadataCache(clientConfig, None, containerConfig, feedRange).block()
+    newItem.feedRange shouldEqual feedRange
+    newItem.lastRetrieved.get should be >= startEpochMs
+    newItem.lastUpdated.get shouldEqual startEpochMs
+    newItem should be theSameInstanceAs testMetadata
+
+    val reinitializedEpochMs = Instant.now.toEpochMilli
+    this.reinitialize()
+
+    val realItem = PartitionMetadataCache(clientConfig, None, containerConfig, feedRange).block()
+    realItem.lastUpdated.get should be >= reinitializedEpochMs
+    realItem should not (be theSameInstanceAs testMetadata)
+  }
+
+  private[this] def reinitialize() = {
+    val container = cosmosClient.getDatabase(cosmosDatabase).getContainer(cosmosContainer)
+    this.feedRange = container.getFeedRanges.block.get(0).toString
```

Comment:
why only the first item in the feed range? is that intentional and will that work against non-emulator endpoint?

Reply:
Yes first partition will work independent of number of physical partitions = and for the remaining code it doesn't matter what the actual ranges are.